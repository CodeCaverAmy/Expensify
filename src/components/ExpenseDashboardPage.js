import React from 'react';

const ExpenseDashboardPage = () => (
    <div>
        This is dashboard componenet
    </div>
); 

export default ExpenseDashboardPage;