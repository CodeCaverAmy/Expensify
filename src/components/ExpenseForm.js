import 'react-dates/initialize';
import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';

// moment is the standard for dates
const now = moment();
console.log(now.format('MMMM Do, YYYY'));

export default class ExpenseForm extends React.Component{
    constructor(props) {
        super(props);

        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(), // timestamp
            calendarFocused: false,
            error: ''
        };
    }
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    };
    onAmountChange = (e) => {
        const amount = e.target.value;
        // validate the amount .. 1) must be a number 2) can only have 2 decimals
        // using REGEX ... ^\d*(\.\d{0,2})?$ -- stars with a digit, followed by unlimited digits, optinal . followed by 0-2 digits, no more ($)
        if (amount.match(/^\d*(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }))
        }
    };
    onDateChange = (createdAt) => {
        this.setState(() => ({ createdAt }));
    }
    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }))
    }
    render() {
        return (
            <div>
                <form>
                    <input 
                        type='text' 
                        placeholder='Description'
                        autoFocus
                        value={this.state.description} // read only value
                        onChange={this.onDescriptionChange}
                    />
                    <input 
                        type='text'
                        placeholder='Amount'
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <textarea
                        placeholder='Add a note for your expense (optional)'
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    >
                    </textarea>
                    <button>
                        Add Expense
                    </button>
                </form>
            </div>
        );
    }
}