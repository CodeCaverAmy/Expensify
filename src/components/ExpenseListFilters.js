import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';
import 'react-dates/initialize';
import { DateRangePicker } from 'react-dates';

// since we need to add state to track the DateChangerPicker, our original stateless component needs to now become  a class based component
export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    };
    onDatesChange = ({ startDate, endDate }) => { // called by the react-dates-libraty with an object with startDate and endDate
        // when the dates change use the actions setStartDate and setEndDate to send to the Redux store
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };
    onFocusChange = (calendarFocused) => { // gets the new value and sets it
        this.setState(() => ({ calendarFocused })); // pass in the updater function by implicitly returning the object calendarFocused: calendarFocues
    };
    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };
    onSortChange = (e) => {
        if (e.target.value === 'date') {
            this.props.sortByDate();
        } else if (e.target.value === 'amount') {
            this.props.sortByAmount();
        }
    };
    render() {
        return (
            <div>
                <input 
                    type='text' 
                    value={this.props.filters.text} 
                    onChange={this.onTextChange}/>
                <select 
                    value={this.props.filters.sortBy} 
                    onChange={this.onSortChange}>
                    <option value='date'>Date</option>
                    <option value='amount'>Amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate}    // required
                    startDateId="startDateFilter"               // required,
                    endDate={this.props.filters.endDate}        // required
                    endDateId="endDateFilter"                   // required,
                    onDatesChange={this.onDatesChange}          // required
                    focusedInput={this.state.calendarFocused}   // required
                    onFocusChange={this.onFocusChange}          // rquired
                    showClearDates={true}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
            </div>
        );
    }
};

const mapStateToProps = (state) => ({
  filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate))
});


export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);