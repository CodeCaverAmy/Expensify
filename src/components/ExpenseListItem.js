import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { removeExpense } from '../actions/expenses';

// Export a stateless functional component
// description, amount, createdAt

const ExpenseListItem = ({ dispatch, id, description, amount, createdAt }) => (
    <div>
        <Link to={`/edit/${id}`}><h3>{description}</h3></Link>
        <p>{amount} - {createdAt}</p>
        <button id={id} onClick={() => {
            dispatch(removeExpense({ id}));
            console.log({ id });
        }}>Remove</button>
    </div>
);

// we don't need anything from state, so we don't need mapStateToProps
export default connect()(ExpenseListItem);