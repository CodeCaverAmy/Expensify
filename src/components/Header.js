import React from 'react';
import { NavLink } from 'react-router-dom';

const Header = () => (
    <header>
        <h1>Expensify</h1>
        <NavLink to='/' exact={true} activeClassName='is-active'>Home</NavLink>
        <NavLink to='/dashboard' activeClassName='is-active'>Dashboard</NavLink>
        <NavLink to='/login' activeClassName='is-active'>Login</NavLink>
    </header>
);

export default Header;