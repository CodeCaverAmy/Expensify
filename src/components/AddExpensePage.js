import React from 'react';
import { connect } from 'react-redux'; // we wand to connect this component to the store so that we can dispatch
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';

// with connect, we now have access to props.dispatch
export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.startAddExpense(expense);
        this.props.history.push('/');
    };
    render() {
        return (
            <div>
                <div className='page-header'>
                    <div className='content-container'>
                        <h1 className='page-header__title'>Add Expense</h1>
                    </div>
                </div>
                <div className='content-container'>
                    <ExpenseForm 
                        onSubmit={this.onSubmit}
                    />
                </div>
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage); // we don't need anything from the state, so the first () can be left empty