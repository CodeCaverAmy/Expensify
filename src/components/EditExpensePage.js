import React from 'react';
import { connect } from 'react-redux'; // we want to connect this component to the redux store
import ExpenseForm from './ExpenseForm'; // we need the expense form so that we can use it in our component
import { editExpense, removeExpense } from '../actions/expenses'; // we need access to the editExpense action

const EditExpensePage = (props) => {
    return (
        <div>
            <ExpenseForm 
                expense={props.expense} // send in the expense into the Expense Form
                onSubmit={(expense) => {
                    // dispatch the action to edit the expense
                    props.dispatch(editExpense(props.expense.id, expense));
                    // redirect to the dashboard page
                    props.history.push('/');
                }}
            />
            <button onClick={() => {
                props.dispatch(removeExpense({id: props.expense.id }));
                // redirect to the dashboard page
                props.history.push('/');
            }}>Remove</button>
        </div>
    );
}; 

// set up mapToStateProps to give the component the current expense object
const mapToStateProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    };
};

export default connect(mapToStateProps)(EditExpensePage); 