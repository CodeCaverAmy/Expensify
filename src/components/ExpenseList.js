import React from 'react';
import { connect } from 'react-redux'; // import named export connect, connecting our component to the Redux store
import ExpenseListItem from './ExpenseListItem';
import SelectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? (
                <p>No expenses</p>

            ) : (
                props.expenses.map((expense) => { 
                    return <ExpenseListItem key={expense.description} {...expense} />
                })            
            )
        }
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: SelectExpenses(state.expenses, state.filters)
    };
};

// create  new const for the HOC
export default connect(mapStateToProps)(ExpenseList); // connect gives something back - a function 

