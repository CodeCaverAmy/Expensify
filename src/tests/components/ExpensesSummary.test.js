import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';

test('should render an expenses summary component with total expense for a singular expense', () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={1} expensesTotal={235}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should render an expenses summary component with total expense for multiple expenses', () => {
    const wrapper = shallow(<ExpensesSummary expenseCount={23} expensesTotal={789522361}/>);
    expect(wrapper).toMatchSnapshot();
});