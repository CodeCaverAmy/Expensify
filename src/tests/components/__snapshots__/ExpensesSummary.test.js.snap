// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render an expenses summary component with total expense for a singular expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totalling 
    $2.35
  </h1>
</div>
`;

exports[`should render an expenses summary component with total expense for multiple expenses 1`] = `
<div>
  <h1>
    Viewing 
    23
     
    expenses
     totalling 
    $7,895,223.61
  </h1>
</div>
`;
